pkg/providers/vsphere/virtualmachine/snapshot.go

```diff
@@ -172,3 +176,161 @@ func (m snapshotMap) add(parent string, tree []types.VirtualMachineSnapshotTree)
 
 	}
 }
+
+// GetSnapshotSize calculates the size of a given snapshot in bytes. It include
+// the memory file, the vmdk files, and the vmsn file.
+// Additionally, it excludes the FCDs.
+// The algorithm use the same logic as how VC UI calculates the snapshot size.
+func GetSnapshotSize(vmCtx pkgctx.VirtualMachineContext, vmSnapshot *types.ManagedObjectReference) int64 {
+	if vmSnapshot == nil || vmSnapshot.Value == "" {
+		vmCtx.Logger.V(5).Info("vmSnapshot is nil or empty")
+		return 0
+	}
+
+	if vmCtx.MoVM.LayoutEx == nil {
+		vmCtx.Logger.V(5).Info("vmCtx.MoVM.LayoutEx is nil, skip calculating snapshot size")
+		return 0
+	}
+	vmlayout := vmCtx.MoVM.LayoutEx
+	vmDevices := vmCtx.MoVM.Config.Hardware.Device
+
+	var fileKeyList []int
+
+	// Find the VirtualMachineFileLayoutExSnapshotLayout for current snapshot
+	for _, snapshot := range vmlayout.Snapshot {
+		if snapshot.Key.Value == vmSnapshot.Value {
+			// Add the memoryKey if it exists.
+			if snapshot.MemoryKey != -1 { // .vmem
+				vmCtx.Logger.V(5).Info("Adding memoryKey", "memoryKey", snapshot.MemoryKey)
+				fileKeyList = append(fileKeyList, int(snapshot.MemoryKey))
+			}
+
+			// .vmsn
+			vmCtx.Logger.V(5).Info("Adding dataKey", "dataKey", snapshot.DataKey)
+			fileKeyList = append(fileKeyList, int(snapshot.DataKey))
+
+			// Add the last element of fileKeys in the disk.chain.
+			// .vmdk files
+			for _, disk := range snapshot.Disk {
+				for _, chain := range disk.Chain {
+					if len(chain.FileKey) > 0 {
+						vmCtx.Logger.V(5).Info("Adding fileKey", "fileKey", chain.FileKey[len(chain.FileKey)-1])
+						fileKeyList = append(fileKeyList, int(chain.FileKey[len(chain.FileKey)-1]))
+					}
+				}
+			}
+		}
+	}
+
+	// Find the keySets for FCDs from layoutEx.file
+	fcdDeviceFileKeySet := getFCDDevicesFileKeySet(vmCtx, vmDevices, vmlayout.File)
```

Comment:
```suggestion
	// A VM snapshot creates a delta disk for all disks of a VM -- including
	// PVCs that are backed by FCDs.  Since the usage of the delta disks 
	// created on PVCs will already be reported by the VolumeSnapshot 
	// SPU, we need to skip those here to avoid double counting. 
```